### PR TITLE
List handler

### PR DESCRIPTION
**Pull request by divishaj**
*Description of changes:* Implementing list handler on AWS::ServiceCatalog::ServiceAction and AWS::ServiceCatalog::ServiceActionAssociation resource.

*Testing:*
pre-commit run --all-files
mvn clean install

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java [hunk 1]
```diff
@@ -79,6 +84,26 @@ public DescribeServiceActionResponse describeServiceAction(final String id) {
...
+                    .pageToken(pageToken)
+                    .build();
+            logger.log(LIST_SERVICE_ACTIONS_LOG);
+            ListServiceActionsResponse response = proxy.injectCredentialsAndInvokeV2(request, scClient::listServiceActions);
+            List<ServiceActionSummary> serviceActionSummaries = response.serviceActionSummaries();
+            pageToken = response.nextPageToken();
+            for (ServiceActionSummary serviceActionSummary: serviceActionSummaries) {
+                serviceActionIds.add(serviceActionSummary.id());
+            }
+        } while(!StringUtils.isNullOrEmpty(pageToken));
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java:102]: Can we benefit from SDK2.0 to have an automated paginator or is this not available to service catalog APIs? See for details: https://aws.amazon.com/blogs/developer/auto-pagination-feature-in-java-sdk-2-0/

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java [hunk 2]
```diff
@@ -108,4 +133,15 @@ static public ResourceModel buildResourceModelFromServiceActionDetail(final Serv
...
+
+    static public List<ResourceModel> buildListResourceModel(List<String> serviceActionIds) {
+        final List<ResourceModel> models = new ArrayList<>();
+        for(String actionId : serviceActionIds){
+            ResourceModel resourceModel = ResourceModel.builder()
+                    .id(actionId)
+                    .build();
+            models.add(resourceModel);
+        }
+        return models;
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java:145]: Use streams api. Something like this: 

```
return serviceActionIds.map(id -> ResourceModel.builder().id(id).build()).collections(toList());
```

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        try {
+            final List<String> serviceActionIds = actionController.listServiceAction();
+            final List<ResourceModel> models = ActionController
+                    .buildListResourceModel(serviceActionIds);
+            return ProgressEvent.<ResourceModel, CallbackContext>builder()
+                    .resourceModels(models)
+                    .status(OperationStatus.SUCCESS)
+                    .build();
+
+        } catch (InvalidParametersException e) {
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java:39]: Are there any other exceptions that must be handled?

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    private static final String LIST_ERROR_MSG = "Unable to list service action because: %s";
+
+    @Override
+    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
+            final AmazonWebServicesClientProxy proxy,
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final Logger logger){
+
+        ActionController actionController = ActionController
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java:24]: nit: Use final in the variables.

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                .builder()
+                .logger(logger)
+                .proxy(proxy)
+                .scClient(SCClientBuilder.getClient())
+                .build();
+        try {
+            final List<String> serviceActionIds = actionController.listServiceAction();
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java:31]: nit: Should the method name be plural? `listServiceActions`?

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            final List<ResourceModel> models = ActionController
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java:32]: Why action controller needs to know how to build the resource models? This is the handler responsibility.

**Review by hugomarques — CHANGES_REQUESTED** [review]
The comments I left in the Service Action handler are also valid for the association handler.

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java (hunk 3 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java:39]: No. There is only one exception thrown by listserviceaction API as mentioned [here.](https://docs.aws.amazon.com/servicecatalog/latest/dg/API_ListServiceActions.html)

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java (hunk 1 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java]: Added automated paginator in new commit

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java (hunk 2 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ActionController.java]: Updated code in new commit

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java (hunk 3 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java]: Added ExceptionTranslator which catch all type of exceptions and throw Cfn exceptions

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java (hunk 4 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java]: Updated code in new commit

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java (hunk 5 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java]: Updated code in new commit

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java (hunk 6 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/ListHandler.java]: Updated code in new commit

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java [hunk 7]
```diff
@@ -13,34 +10,27 @@
...
     @Override
     public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
-        final AmazonWebServicesClientProxy proxy,
-        final ResourceHandlerRequest<ResourceModel> request,
-        final CallbackContext callbackContext,
-        final Logger logger){
+            final AmazonWebServicesClientProxy proxy,
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final Logger logger){
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java:18]: Why are we breaking identation here?

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java [hunk 8]
```diff
@@ -13,34 +10,27 @@
...
-        } catch (InvalidParametersException e) {
-            throw new CfnInvalidRequestException(String.format(CREATE_ERROR_MSG, e.getMessage()), e);
-
-        } catch (LimitExceededException e) {
-            throw new CfnServiceLimitExceededException(ResourceModel.TYPE_NAME, e.getMessage(), e);
+            desiredModel = ActionController.buildResourceModelFromServiceActionDetail(response.serviceActionDetail());
+        } catch (SdkException e) {
+            ExceptionTranslator.translateToCfnException(e);
         }
+        return ProgressEvent.defaultSuccessHandler(desiredModel);
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java:34]: Why is the return outside of the try catch?

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java [hunk 9]
```diff
@@ -13,34 +10,27 @@
...
+            final CallbackContext callbackContext,
+            final Logger logger){
 
         ActionController actionController = ActionController
                 .builder()
                 .logger(logger)
                 .proxy(proxy)
                 .scClient(SCClientBuilder.getClient())
                 .build();
-        final ResourceModel desiredModel = request.getDesiredResourceState();
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java:31]: If return is inside try catch there's no need to remove final.

File: aws-servicecatalog-serviceaction/aws-servicecatalog-serviceaction.json [hunk 10]
```diff
@@ -94,6 +94,11 @@
       "permissions": [
         "servicecatalog:DeleteServiceAction"
       ]
+    },
+    "list": {
+      "permissions": [
+        "servicecatalog:ListServiceActionsPaginator"
```
Comment (hugomarques) [inline aws-servicecatalog-serviceaction/aws-servicecatalog-serviceaction.json:100]: Is this right? Do we have to specifify the paginator? What about the FAS policy?

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java (hunk 7 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java]: removed in new revision

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java (hunk 9 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java]: Updated code

File: aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java (hunk 8 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/src/main/java/software/amazon/servicecatalog/serviceaction/CreateHandler.java]: Updated code

File: aws-servicecatalog-serviceaction/aws-servicecatalog-serviceaction.json (hunk 10 above)
Comment (divishaj) [reply aws-servicecatalog-serviceaction/aws-servicecatalog-serviceaction.json]: Updated permission

**Review by hugomarques — COMMENTED** [review]
Did we run integration tests to make sure the permissions in the schema are right?

**Review by hugomarques — APPROVED** [review]

**Review by mrmayank-amazon — APPROVED** [review]

**Pull request merged**